### ISSUE TITLE
Tracking progress in implementation of Dynamic Graph struct

### ISSUE THREAD

**Issue by wolfram77** [puzzlef/graph-openmp#3]
I just implemented a concurrent arena allocator, which given a memory pool, returns allocated memory of fixed size memory blocks. It also maintains a list of freed memory blocks, and any new requests are serviced from the freed blocks.

However, in order to be thread-safe, i.e., to support memory allocation calls from multiple threads concurrently, we are using a `atomic_flag` which is used as a mutex to limit access to the freed list to one thread at a time only. In order to ensure that no thread has to wait for too long, we check if the `flag` is set to `busy`, and if so, we go ahead an return a block from the pool (which uses an atomic add). However, it is also possible that we have no free space in the pool. In such a case, we repeatedly retry acquiring the mutex (`yield`ing the current thread on failure and retrying), and once obtained, fetch a memory block from the freed list.

For freeing a memory block, however, we must access the memory pool, so we we just do repeated retries to access the freed list, and then, once acquired, append the memory block to the freed list.

The results look like the following, when (several) allocations are being performed sequentially

```bash
- malloc(): ~10s
- arena.alloc(): ~5s

- free(): ~2s
- arena.free(): ~1s
```

However, when allocations are done in parallel using 64 thread, the results look quite different

```bash
- malloc(): ~20s
- arena.alloc(): ~20s

- free(): ~1s
- arena.free(): too long
```

It appears that there is high contention in our concurrent arena allocator. Sad! We could try using per-thread freed lists to resolve this. But then, we would have to implement some stealing mechanism in order to fetch memory blocks that have been freed by other threads. Might as well use **libc** `malloc()` instead - particularly for large size memory allocations.

In fact, we have another recursive arena allocator which utilizes multiple pools to allocate smaller memory blocks to requesting thread, and is not thread-safe (for one thread only). A suitable way, then, might be to use `malloc()` for large allocations, and revert to per-thread recursive arena allocator for smaller allocations - we may now called it multi arena allocator instead. LGTM.

See #4 for the code.

**Comment by wolfram77** [comment]
Let's think what a graph would require. Concurrent allocators of various sizes.
- Quickly free and allocate a new memory block. It also better be multi-level.
- Ensure that vertex access is efficient, by using SoA vectors.
- Ensure that edge weights are stored in a separate array.
- Delay merging of edges into a single array, until the graph is finalized.
- Select a different merging startegy, based on the number of edges inserted / deleted.
- Perform a fast deletion of edges, i.e., deletions come first.

**Comment by wolfram77** [comment]
Measuring performance of CSR read vs Graph read:

![image](https://github.com/user-attachments/assets/7eeaf592-89e0-4656-9c64-4458a737cf15)

**Comment by wolfram77** [comment]
> Now able to handle symmetric graphs: 
> [output-graph-openmp--load-undirected-with-double-edges-in-edgelist.log](https://gist.github.com/wolfram77/9f98128160a7b4e525e4c999a65c7af7#file-output-graph-openmp-load-undirected-with-double-edges-in-edgelist-log)

**Comment by wolfram77** [comment]
> Now able to convert CSR to `DiGraph`: \
> [output-graph-openmp--convert-csr-to-digraph.log](https://gist.github.com/wolfram77/6ef2a7fa2a423d2acf09c77788f458fe#file-output-graph-openmp-convert-csr-to-digraph-log)

**Comment by wolfram77** [comment]
Interesting to see the runtime split of `duplicateIfOmpW()`, which is used to convert the graph from CSR to DiGraph. Reserving space for the graph, and for the edges appear to be the largest bottlenecks, followed by `updateOmpU()`.

> Logs: [output-graph-openmp--profile-duplicate-graph.log](https://gist.github.com/wolfram77/887dedce5fd61650083ebfa09c339fef#file-output-graph-openmp-profile-duplicate-graph-log)

![image](https://github.com/user-attachments/assets/e6a85df7-034f-4c22-b2f9-70dc5d0344d3)

**Comment by wolfram77** [comment]
Currently attempting to load into arena graph, but have hit the wall with a double free error during reserving memory for the edges.

![Image](https://github.com/user-attachments/assets/cfe5a0c9-fe16-4f9c-be08-8745b35b2002)

The relevant code is given below:

```cxx
  #pragma omp parallel for schedule(dynamic, 2048)
  for (K u=0; u<S; ++u) {
    if (!x.hasVertex(u)) continue;
    a.allocateEdges(u, x.degree(u));
  }
```

Arena-based digraph is now working. The double-free issue was due to using a single allocator across all threads. Another issue we faced was the failure to populate the edges, resulting in abrupt program crash. This was due to using `__builtin_clz()` for computing capacity of the slots, but `__builtin_clz()` only works for 32-bit integer, not `size_t`. Below are the results after fixing.

![Image](https://github.com/user-attachments/assets/60097138-c6f9-4f70-9392-6cedafa775e7)

**Comment by wolfram77** [comment]
> Upto 2KB are handled by arena allocator (with a capacity of 8 * 64KB), remaining are managed by libc. \
> [output-graph-openmp--approach-csr-graph-no-sort.log](https://gist.github.com/wolfram77/963282c413553d2a4518758593e6f8da)

![Image](https://github.com/user-attachments/assets/215916e2-8c19-4572-b95e-79d469c75017)

![Image](https://github.com/user-attachments/assets/c658207f-d9b6-4623-8f36-072151032fb6)

**Comment by wolfram77** [comment]
> Upto 8KB are handled by arena allocator (with a capacity of 10 * 512KB), remaining are managed by libc. \
> BTW, why is resizing the arrays for vertices too slow, particularly for kmer graphs? \
> [output-graph-openmp--approach-duplicate-8192.log](https://gist.github.com/wolfram77/44bad94e26f40282229ff0530309b8ed)

![Image](https://github.com/user-attachments/assets/b9f17c23-7cef-4803-8ccf-9c7ea58c8f96)

![Image](https://github.com/user-attachments/assets/5e3ff858-5e8d-442a-8a2d-92dcc4cc6393)

> [!NOTE]
> We also try to minimize the memory being held by the arena allocators. Tested on sk-2005 graph. Best config seems to be using up to 8KB arena allocators, with capacity of each allocator being 512KB.

**Comment by wolfram77** [comment]
> Upto 8KB are handled by arena allocator (with a capacity of 10 * 512KB), remaining are managed by libc \
> Now we are managing the vertex arrays on our own, so we can initialize them in parallel \
> The improved locality of the variables seems to be improving the performance of add-edges as well. \
> [output-graph-openmp--approach-duplicate-8192-custom-arrays.log](https://gist.github.com/wolfram77/abebd1e128a58225093a027589facfb0)

![Image](https://github.com/user-attachments/assets/5adb6608-c5ae-4755-8b11-be40caa275d7)

![Image](https://github.com/user-attachments/assets/4e1cea51-bf66-49d7-95e5-5f60b660dd00)

> [!NOTE]
> `vector.resize()` is sequential and is a bottleneck for graphs with a large number of vertices (such as k-mer graphs).

**Comment by wolfram77** [comment]
> It seems choosing a suitable chunk size has a significant impact on sorting edges.
> [output-graph-openmp--measure-sort-time.log](https://gist.github.com/wolfram77/3e87939ca44ddb5f2cd878be1293db21)

![Image](https://github.com/user-attachments/assets/fe0a77db-6235-4a84-8ab9-87cef4f2e11b)

![Image](https://github.com/user-attachments/assets/189f8be4-217f-4810-9d38-c66d0577e68b)

**Comment by wolfram77** [comment]
> Dynamic schedule with task `(deg > 2048)` seems best. \
> [output-graph-openmp--adjust-sort-method.log](https://gist.github.com/wolfram77/09091de040c47ec8b846dc1c8d58e373)

![Image](https://github.com/user-attachments/assets/98ae5d34-5c6e-4abd-9a40-c7f7e44a2e1e)

**Comment by wolfram77** [comment]
> Below is the phase split of duplicate graph (arena allocator-based). \
> [puzzlef/graph-openmp--measure-duplicate-split](https://docs.google.com/spreadsheets/d/1u57UJt01fpc6So5lPRzaf7STeYfN4l-c1caE0fBbLwY/edit?usp=sharing)

![Image](https://github.com/user-attachments/assets/52fbe9df-1cfc-4eac-a089-26db4d752be2)

**Comment by wolfram77** [comment]
> Measuring runtime with various chunk size for add and subtract graph operations. \
> [output-graph-openmp--adjust-add-subtract-chunk-size.log](https://gist.github.com/wolfram77/14d7d391a5e70407389844f24df78673)

**Comment by wolfram77** [comment]
> Results with varying batch fraction. \
> [output-graph-openmp--adjust-batch-fraction.log](https://gist.github.com/wolfram77/ff226c8cbbc4decc326293246d2a5196)

![Image](https://github.com/user-attachments/assets/c1c45e4b-e595-404b-a21d-210c2292e04a)

**Comment by wolfram77** [comment]
> Running all experiments: \
> [output-graph-openmp--all-expts.log](https://gist.github.com/wolfram77/d183684f9deab096b53ac5b52d352e8a)

**Comment by wolfram77** [comment]
Results are below:

![Image](https://github.com/user-attachments/assets/3ef09c02-3fa1-4558-9cef-45531093a47a)

![Image](https://github.com/user-attachments/assets/7252a1fc-abb2-4f38-afd5-0479e08c6468)

![Image](https://github.com/user-attachments/assets/a5edbd2e-012e-4d52-9edc-fc5f83bf2d21)

![Image](https://github.com/user-attachments/assets/60ec8b22-e4d0-4ee7-8752-21bda26add01)
